File: mac/BG3Assistant/PetAnimationModel.swift
import CoreGraphics
import Foundation

struct PetSpriteFrame: Equatable {
    let row: Int
    let column: Int
}

enum PetAnimationModel {
    // Matches the authored Codex v2 atlas timings.
    static let idleDurations = [280, 110, 110, 140, 140, 320]
    static let jumpingDurations = [140, 140, 140, 140, 280]

    static var hoverIntroDuration: TimeInterval {
        TimeInterval(jumpingDurations.reduce(0, +)) / 1_000
    }

    static func frame(
        isHovered: Bool,
        hoverElapsed: TimeInterval,
        pointerLocation: CGPoint?,
        viewSize: CGSize,
        reduceMotion: Bool
    ) -> PetSpriteFrame {
        guard isHovered else { return PetSpriteFrame(row: 0, column: 0) }

        if !reduceMotion, hoverElapsed < hoverIntroDuration {
            return PetSpriteFrame(
                row: 4,
                column: animationColumn(elapsed: hoverElapsed, durations: jumpingDurations)
            )
        }

        if let pointerLocation,
           let lookFrame = lookFrame(pointerLocation: pointerLocation, viewSize: viewSize) {
            return lookFrame
        }

        guard !reduceMotion else { return PetSpriteFrame(row: 0, column: 0) }
        return PetSpriteFrame(
            row: 0,
            column: animationColumn(
                elapsed: max(0, hoverElapsed - hoverIntroDuration),
                durations: idleDurations
            )
        )
    }

    static func lookFrame(pointerLocation: CGPoint, viewSize: CGSize) -> PetSpriteFrame? {
        let center = CGPoint(x: viewSize.width / 2, y: viewSize.height / 2)
        let dx = pointerLocation.x - center.x
        let dy = pointerLocation.y - center.y
        let deadzone = min(viewSize.width, viewSize.height) * 0.14
        guard hypot(dx, dy) > deadzone else { return nil }

        // SwiftUI's y axis points down. atan2(dx, -dy) makes 000 point up
        // and advances clockwise, matching v2 rows 9 and 10.
        var degrees = atan2(dx, -dy) * 180 / .pi
        if degrees < 0 { degrees += 360 }
        let directionIndex = Int(floor((degrees + 11.25) / 22.5)) % 16
        return PetSpriteFrame(
            row: directionIndex < 8 ? 9 : 10,
            column: directionIndex % 8
        )
    }

    private static func animationColumn(elapsed: TimeInterval, durations: [Int]) -> Int {
        let total = durations.reduce(0, +)
        guard total > 0 else { return 0 }
        var remaining = Int(max(0, elapsed) * 1_000) % total
        for (column, duration) in durations.enumerated() {
            if remaining < duration { return column }
            remaining -= duration
        }
        return max(0, durations.count - 1)
    }
}

File: windows/src/BG3HonorAssistant.Core/Overlay/PetAnimationModel.cs
namespace BG3HonorAssistant.Core.Overlay;

public readonly record struct PetSpriteFrame(int Row, int Column);

public readonly record struct PetPoint(double X, double Y);

public readonly record struct PetSize(double Width, double Height);

public static class PetAnimationModel
{
    public static readonly IReadOnlyList<int> IdleDurations =
        [280, 110, 110, 140, 140, 320];

    public static readonly IReadOnlyList<int> JumpingDurations =
        [140, 140, 140, 140, 280];

    public static TimeSpan HoverIntroDuration =>
        TimeSpan.FromMilliseconds(JumpingDurations.Sum());

    public static PetSpriteFrame Frame(
        bool isHovered,
        TimeSpan hoverElapsed,
        PetPoint? pointerLocation,
        PetSize viewSize,
        bool reduceMotion)
    {
        if (!isHovered)
        {
            return new PetSpriteFrame(0, 0);
        }

        if (!reduceMotion && hoverElapsed < HoverIntroDuration)
        {
            return new PetSpriteFrame(
                4,
                AnimationColumn(hoverElapsed, JumpingDurations));
        }

        if (pointerLocation is { } pointer &&
            LookFrame(pointer, viewSize) is { } look)
        {
            return look;
        }

        if (reduceMotion)
        {
            return new PetSpriteFrame(0, 0);
        }

        return new PetSpriteFrame(
            0,
            AnimationColumn(
                hoverElapsed - HoverIntroDuration < TimeSpan.Zero
                    ? TimeSpan.Zero
                    : hoverElapsed - HoverIntroDuration,
                IdleDurations));
    }

    public static PetSpriteFrame? LookFrame(
        PetPoint pointerLocation,
        PetSize viewSize)
    {
        var centerX = viewSize.Width / 2D;
        var centerY = viewSize.Height / 2D;
        var dx = pointerLocation.X - centerX;
        var dy = pointerLocation.Y - centerY;
        var deadzone = Math.Min(viewSize.Width, viewSize.Height) * 0.14D;
        if (Math.Sqrt((dx * dx) + (dy * dy)) <= deadzone)
        {
            return null;
        }

        var degrees = Math.Atan2(dx, -dy) * 180D / Math.PI;
        if (degrees < 0D)
        {
            degrees += 360D;
        }

        var direction = (int)Math.Floor((degrees + 11.25D) / 22.5D) % 16;
        return new PetSpriteFrame(
            direction < 8 ? 9 : 10,
            direction % 8);
    }

    private static int AnimationColumn(
        TimeSpan elapsed,
        IReadOnlyList<int> durations)
    {
        var total = durations.Sum();
        if (total <= 0)
        {
            return 0;
        }

        var remaining = (int)Math.Max(0D, elapsed.TotalMilliseconds) % total;
        for (var column = 0; column < durations.Count; column++)
        {
            if (remaining < durations[column])
            {
                return column;
            }

            remaining -= durations[column];
        }

        return Math.Max(0, durations.Count - 1);
    }
}
